tesseract_core/sdk/logs.py

```diff
@@ -25,6 +27,59 @@
 }
 
 
+# NOTE: This is duplicated in `tesseract_core/runtime/logs.py`.
+# Make sure to propagate changes to both files.
+class LogPipe(threading.Thread):
+    """Custom IO pipe to support live logging from subprocess.run or OS-level file descriptor.
+
+    Runs a thread that logs everything read from the pipe to the given sinks.
+    Can be used as a context manager for automatic cleanup.
+    """
+
+    daemon = True
+
+    def __init__(self, *sinks: Callable) -> None:
+        """Initialize the LogPipe with the given logging level."""
+        super().__init__()
+        self._sinks = sinks
+        self._fd_read, self._fd_write = os.pipe()
+        self._pipe_reader = os.fdopen(self._fd_read)
+        self._captured_lines = []
+
+    def __enter__(self) -> int:
+        """Start the thread and return the write file descriptor of the pipe."""
+        self.start()
+        return self.fileno()
+
+    def __exit__(self, *args: Any) -> None:
+        """Close the pipe and join the thread."""
+        os.close(self._fd_write)
+        # Use a timeout so something weird happening in the logging thread doesn't
+        # cause this to hang indefinitely
+        self.join(timeout=10)
+        # Do not close reader before thread is joined since there may be pending data
+        # This also closes the fd_read pipe
+        self._pipe_reader.close()
+
+    def fileno(self) -> int:
+        """Return the write file descriptor of the pipe."""
+        return self._fd_write
+
+    def run(self) -> None:
+        """Run the thread, logging everything."""
+        for line in iter(self._pipe_reader.readline, ""):
```

Comment:
On my branch I had better success with `shutil.copyfilobj`, which works much better than a line-based approach for e.g. tqdm progress bars.

Reply:
Yea I can see how the line-based parsing can cause problems but I believe this is the best we can do for now. `copyfileobj` is creating a 1-1 map but we're mapping one fd source to two sinks.